numpy/core/src/multiarray/compiled_base.c

```diff
@@ -795,7 +795,52 @@ static int sequence_to_arrays(PyObject *seq,
     return 0;
 }
 
-/* Inner loop for unravel_index */
+/*
+ * Force-cast an array into intp if it is empty. Fail otherwise.
+ *
+ * op        - Pointer to the array pointer. Will be modified if
+ *             recasting is needed.
+ * paramname - Name to be used for error string.
+ */
+static int _cast_to_intp_or_fail(PyArrayObject **op,
+                                 char *paramname)
+{
+    PyArray_Descr *dtype = PyArray_DescrFromType(NPY_INTP);
+    if (dtype == NULL) {
+        return -1;
+    }
+
+    if (!PyArray_CanCastArrayTo(*op, dtype, NPY_SAME_KIND_CASTING)) {
```

Comment:
You are not casting arrays that can be casted, only empty ones!

Here's how I would write this, feel free to draw inspiration from it:

```C
static int
_cast_to_intp_or_fail(PyArrayObject **op, char *paramname)
{
    PyArray_Descr *dtype = PyArray_DescrFromType(NPY_INTP);

    if (PyArray_CanCastArrayTo(*op, dtype, NPY_SAME_KIND_CASTING) ||
            PyArray_SIZE(*op) == 0) {
        PyObject *op_tmp = PyArray_FromArray(*op, dtype, NPY_ARRAY_FORCECAST);

        if (op_tmp == NULL) {
          return -1;
        }

        Py_DECREF(*op);
        *op = (PyArrayObject *)op_tmp;

        return 0;
    }
    
    Py_DECREF(dtype);
    PyErr_Format(PyExc_TypeError,
                 "non-empty %s array has a non-integer dtype", paramname);
    
    return -1;
}
```